### ISSUE TITLE
[BUG] Server hangs on changelevel

### ISSUE THREAD

**Issue by blobles-dev** [FredyH/MySQLOO#138]
My server is suffering from a weird issue where it just hangs when changelevel occurs, no actual crash, just unresponsive. Some research has led me to this module being the cause..

This is related to: #118

Big shoutout to Warden Potato on the gmod discord for making me aware and providing this code to replicate:

```
local DB = mysqloo.Connect(host, username, password, database)

function DebuggingBusyQuery()
    local q = DB:query("SELECT * FROM `sometable` WHERE `somevalue` = 'SOMETHING' ORDER BY `order` DESC LIMIT " .. math.random(1, 10000))
    function q:onSuccess(rows)
    	--optional print
        --print("Result from busy query")
    end
    function q:onError() print("Errored on debug query") end
    q:start()
end

concommand.Add("robin_williams", function(ply)
	if not ply:IsSuperAdmin() then return end
	timer.Simple(2, function()
		for i = 1, 15000, 1 do
			DebuggingBusyQuery()
		end
		
		RunConsoleCommand("changelevel", "gm_flatgrass")
		print("changelevel has begun")
		
		for i = 1, 8000, 1 do
			DebuggingBusyQuery()
		end
	end)
end)
```

NOTE: The db should be remote, and the query should be somewhat slow, this ensures issue replication.

A simple workaround is:

```
hook.Add("ShutDown", "MySQLOO:KillQueriesOnShutdown", function()
	ServerLog("Shutting down a active DB connection and killing all running queries (".. DB:queueSize() ..")\n")

	DB:abortAllQueries()
	DB:disconnect(false)
end)
```

However, this could result in data loss, plus, you'd have to specifically add an instance for each DB, unless there's a way to grab all DB objects.

**Comment by FredyH** [comment]
The database usually waits until all queries are finished, which is usually something you'd want.

Does the problem you mention only occur if queries are launched after the change level has been run? Or is there a particular reason why the busy queries are split into two parts?

**Comment by blobles-dev** [comment]
> The database usually waits until all queries are finished, which is usually something you'd want.
> 
> Does the problem you mention only occur if queries are launched after the change level has been run? Or is there a particular reason why the busy queries are split into two parts?

I've noticed these methods in the mysqloo table:
["allocationCount"]     =       function: 0xf1baad3a
["deallocationCount"]   =       function: 0xf1baad62
["objectCount"] =       function: 0xf1baacfa
["referenceCreatedCount"]       =       function: 0xf1baadf2
["referenceFreedCount"] =       function: 0xf1baadaa

Would these be beneficial to add to my debug prints prior to my work-around so you can get an idea of perhaps where it is stuck?

**Comment by blobles-dev** [comment]
I have added the following code to log some data:

```lua
        local data = {}
        data["time"] = os.date("%c")
        data["timestamp"] = os.time()
        data["queueSize"] = Prometheus.DB.Module.DB:queueSize()
        data["preAbort"] = {
            allocationCount = mysqloo.allocationCount(),
            deallocationCount = mysqloo.deallocationCount(),
            objectCount = mysqloo.objectCount(),
            referenceCreatedCount = mysqloo.referenceCreatedCount(),
            referenceFreedCount = mysqloo.referenceFreedCount(),
        }
        
        Prometheus.DB.Module.DB:abortAllQueries()

        data["postAbort"] = {
            allocationCount = mysqloo.allocationCount(),
            deallocationCount = mysqloo.deallocationCount(),
            objectCount = mysqloo.objectCount(),
            referenceCreatedCount = mysqloo.referenceCreatedCount(),
            referenceFreedCount = mysqloo.referenceFreedCount(),
        }

        local jsonString = util.TableToJSON(data, true) 
        local fileName = "prometheus_mysqloo_log.txt"
        if file.Exists(fileName, "DATA") then
            file.Append(fileName, "\n"..jsonString)
        else
            file.Write(fileName, jsonString)
        end
        
        Prometheus.DB.Module.DB:disconnect(false)
```

I'll check back in a few days if there's anything that sticks out.

**Comment by blobles-dev** [comment]
@FredyH 
Attached my log file, the last entry is a suspected crash (i had my workaround enabled so cant confirm but all aligns)
I remember when the crashes where occurring I was also getting connection errors from my DB, perhaps its something to do with queries being "stuck" with  disconnected DB? Thats also how I simulated this. Had my DB connected, took its perms so it failed, then waited.
[prometheus_mysqloo_log.txt](https://github.com/user-attachments/files/16806156/prometheus_mysqloo_log.txt)

**Comment by FredyH** [comment]
Yeah so it looks like there are still some queries running.
The problem is that queries can lose the TCP connection and not notice that the server is gone, which will keep them that way indefinitely.
I added timeout functions that you can use to configure when a query should timeout, but I have not released the update yet. I will probably do so later this week.

**Comment by blobles-dev** [comment]
Great news! I saw that commit and figured it probably be related. Two quick questions.. will it call the fail callback and is there a default timeout? It isnt my code thats failing and would be great if I could avoid editing

**Comment by FredyH** [comment]
Hi, I released a beta version with the new connection timeouts.
Unfortunately, there is not a default value (except for the default the mysql connector supplies), so you will have to set it yourself.
However, you could simply override the MySQLOO metatable function for connecting to the database and call the timeout functions beforehand.

**Comment by Zulbem** [comment]
> The problem is that queries can lose the TCP connection and not notice that the server is gone, which will keep them that way indefinitely

Yes, this is exactly what i've been experiencing for some years now, server hangs on timeout after attempting to change maps due to a no TCP connection. Although the lost connection for me has always started straight from the previous map change, all the way through to the upcoming map change (which it never gets to do due to the infinite loop you speak of I think). I will try the beta version now with a reasonable timeout, even though that means data loss :/

**Comment by FredyH** [comment]
> I will try the beta version now with a reasonable timeout

Please let me know if it worked for you!

**Comment by Zulbem** [comment]
> > I will try the beta version now with a reasonable timeout
> 
> Please let me know if it worked for you!

Okay so seems to fix it completely hanging the server, however it still pops up with timeout message for 5-10 seconds upon a map change before it actually changes. This shouldn't happen if I close the database object before shutdown right?

**Comment by FredyH** [comment]
I will close this issue for now as there is not much more I can do from my side.
If people using a different host than Physgun (or anyone using Path) is getting the same issue, please let me know.
